src/S3/Transfer.php

```diff
@@ -211,74 +215,48 @@ private function normalizePath($path)
 
     private function createDownloadPromise()
     {
-        // Prepare args for ListObjects.
-        $listArgs = $this->getS3Args($this->source['path']);
-        if (isset($listArgs['Key'])) {
-            $listArgs['Prefix'] = $listArgs['Key'] . '/';
-            unset($listArgs['Key']);
+        $parts = $this->getS3Args($this->sourceMetadata['path']);
+        $prefix = "s3://{$parts['Bucket']}/"
+            . (isset($parts['Key']) ? $parts['Key'] . '/' : '');
+
+        $commands = [];
+        foreach ($this->getDownloadsIterator() as $object) {
+            // Prepare the sink.
+            $localKey = $object;
+            if ($prefix && strpos($localKey, $prefix) === 0) {
+                $localKey = substr($localKey, strlen($prefix));
+            }
+            $sink = $this->destination['path'] . '/' . $localKey;
+
+            // Create the directory if needed.
+            $dir = dirname($sink);
+            if (!is_dir($dir) && !mkdir($dir, 0777, true)) {
+                return Promise\rejection_for(
+                    new \RuntimeException("Could not create dir: {$dir}")
+                );
+            }
+
+            // Create the command.
+            $commands[] = $this->client->getCommand(
```

Comment:
I don't remember why I didn't do it this way before, but what about creating a generator that yields these build up commands rather than storing them all in an array? Given that we're downloading based on listObjects, the number of GetObject commands could be practically infinite.